### PR TITLE
feature-28 adds changes to database transforming the incident and inc…

### PR DESCRIPTION
**Pull request by Kentus**
…ident type tables from MySam to InnoDB.

adds a new column to incident types table, "status" that can be Active or Inactive, setting the default on Active.
adds a unique index for code on incident types table.
adds a foreign key between incidents and incident types table to prevent incident type to be deleted when there are incidents on that type.
adds CRUD endpoints in api_routes for incident types.
adds the CRUD methods(store, update, show, destroy) in IncidentTypeController using for update and store a service to validate and hydrate the data.
adds a new transformer for Incident Type.

### CODE REVIEW

File: app/Services/IncidentTypeService.php [hunk 1]
```diff
@@ -0,0 +1,76 @@
+<?php
+
+namespace App\Services;
+
+use App\IncidentType;
+use Dingo\Api\Exception\StoreResourceFailedException;
+use Illuminate\Support\Facades\Validator;
+
+class IncidentTypeService
```
Comment (claudiunicolaa) [inline app/Services/IncidentTypeService.php:9]: 👍

File: app/Http/api_routes.php [hunk 2]
```diff
@@ -22,6 +22,11 @@
         $api->resource('page', 'App\Api\V1\Controllers\PageController');
         $api->post('page/{id}/approve', 'App\Api\V1\Controllers\PageController@approve');
         $api->post('page/{id}/reject', 'App\Api\V1\Controllers\PageController@reject');
+        //Incident types routes
+        $api->post('incidents/types/', 'App\Api\V1\Controllers\IncidentTypeController@store');
```
Comment (claudiunicolaa) [inline app/Http/api_routes.php]: take a look to the possibility to use `$api->resource('incidents/types')` instead of defining every route separated.

Comment (Kentus) [reply app/Http/api_routes.php]: I have changed to $api->resource restricting the methods that need the admin role.

**Comment by claudiunicolaa** [comment]
:twisted_rightwards_arrows: From my point of view can be merged.

File: app/Api/V1/Controllers/IncidentTypeController.php [hunk 3]
```diff
@@ -48,4 +64,271 @@ public function index()
...
+     * @param IncidentTypeService $incidentTypeService
+     * @return \Dingo\Api\Http\Response
+     */
+    public function store(Request $request, IncidentTypeService $incidentTypeService)
+    {
+        try {
+            /** @var IncidentType $incidentType */
+            $incidentType = $incidentTypeService->create($request->all());
+        } catch (StoreResourceFailedException $exception) {
+            throw $exception;
```
Comment (RaduCStefanescu) [inline app/Api/V1/Controllers/IncidentTypeController.php:127]: Is there a global exception handler for controllers where this exception is processed?
If there isn't, I would rather return a response with an http internal server error or bad request pr conflict (if the exception is caused by violating a primary key exception) code rather then propagate the exception to the calling environment.

File: app/Http/api_routes.php [hunk 4]
```diff
@@ -22,6 +22,8 @@
         $api->resource('page', 'App\Api\V1\Controllers\PageController');
         $api->post('page/{id}/approve', 'App\Api\V1\Controllers\PageController@approve');
         $api->post('page/{id}/reject', 'App\Api\V1\Controllers\PageController@reject');
+        //Incident types routes
+        $api->resource('incidents/types', 'App\Api\V1\Controllers\IncidentTypeController', ['only' => ['show', 'update', 'store', 'destroy']]);
```
Comment (RaduCStefanescu) [inline app/Http/api_routes.php:26]: One question here.
My impression is that this sets the endpoints as managed by the Route api, but does not set the operations permitted on each endpoint (POST, GET, etc)

File: app/Services/IncidentTypeService.php [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        $incidentType->save();
+
+        return $incidentType;
+    }
+
+    /**
+     * @param array $data
+     * @param null $id
+     */
+    public function validate(array $data, $id = null)
```
Comment (RaduCStefanescu) [inline app/Services/IncidentTypeService.php:46]: Was there a reason you added the id as part of the method signature? I don't see it used anywhere.

**Comment by RaduCStefanescu** [comment]
Hi @Kentus great pull request! made some small comments, can you please have a look on them? Thank you!

File: app/Http/api_routes.php (hunk 4 above)
Comment (Kentus) [reply app/Http/api_routes.php:26]: From Laravel's documentation(https://laravel.com/docs/5.2/controllers), the restrictions are these:
Verb | Path | Action | Route Name
POST | /photo | store | photo.store
GET | /photo/{photo} | show | photo.show
PUT/PATCH | /photo/{photo} | update | photo.update
DELETE | /photo/{photo} | destroy | photo.destroy

File: app/Services/IncidentTypeService.php (hunk 5 above)
Comment (Kentus) [reply app/Services/IncidentTypeService.php:46]: It is used in $rules, 'code' => 'required|unique:incident_types,code,' . $id, more exactly when you want to update an existing object you will get a unique constraint error if you don't pass the current id.

File: app/Api/V1/Controllers/IncidentTypeController.php (hunk 3 above)
Comment (Kentus) [reply app/Api/V1/Controllers/IncidentTypeController.php:127]: There is the standard one, app/Exceptions/Handler.php and I tried to keep the controller appropriate with the other controllers, for eg. PrecinctController.php:85
`if ($validator->fails()) {
            throw new StoreResourceFailedException('Could not create new precinct.', $validator->errors());
  }`.
I can't see how to generate a primary key exception but for the unique exception, the message is pretty generic and it doesn't return any details about the paths or database information. This is handled in the IncidentTypeService::validate():57 and it returns only the fields and their violations( in the standard way of Laravel)
`if ($validator->fails()) {
            throw new StoreResourceFailedException('Could not create a new incident type.', $validator->errors());
  }`

Comment (RaduCStefanescu) [reply app/Api/V1/Controllers/IncidentTypeController.php:127]: Cool, thanks for the explanation.

File: app/Http/api_routes.php (hunk 4 above)
Comment (RaduCStefanescu) [reply app/Http/api_routes.php:26]: Got it, thanks!

**Pull request merged**